### PR TITLE
Update Dockerfile.hsm to use debian:bullseye base image

### LINKED ISSUE #1286: Use glibc OS for HSM images
### Descriptions

The current docker image with CGO support, [Dockerfile.hsm](https://github.com/smallstep/certificates/blob/master/docker/Dockerfile.hsm) is based on alpine, which uses musl instead of glibc. We should use a glibc-based OS instead because proprietary HSM products produce private libraries compatible with glibc-based Linux OSs instead of musl. See https://github.com/smallstep/step-kms-plugin/issues/28#issuecomment-1433909181.

### PR DESCRIPTION
**Pull request by tashian**
- [x] update Docker.hsm to use Debian
- [x] Deprecate the `step-ca-hsm` image in favor of `step-ca:hsm`

To merge & release this PR:

- [ ] Publish cli "bullseye" image _before_ merging this
- [ ] Publish step-kms-plugin "bullseye" image _before_ merging this
- [ ] Publish step-kms-plugin "cloud" image tag _before_ merging this
- [ ] Post-release: Update Docker Hub readmes to explain new tags

Related:
- https://github.com/smallstep/step-kms-plugin/pull/47
- https://github.com/smallstep/cli/pull/861

fixes #1286

### CODE REVIEW

File: docker/Dockerfile.hsm [hunk 1]
```diff
@@ -1,23 +1,25 @@
...
+RUN apt-get install -y --no-install-recommends \
+        gcc pkgconf libpcsclite-dev
 RUN make V=1 download
 RUN make V=1 GOFLAGS="" build
 
-FROM smallstep/step-kms-plugin:latest AS kms
+FROM smallstep/step-kms-plugin:debian AS kms
 
-FROM smallstep/step-cli:latest
+FROM smallstep/step-cli:debian
```
Comment (maraino) [inline docker/Dockerfile.hsm:14]: You mentioned that you were going to use the tag `:bullseye`

File: docker/Dockerfile.hsm [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
-FROM smallstep/step-cli:latest
+FROM smallstep/step-cli:bullseye
 
 COPY --from=builder /src/bin/step-ca /usr/local/bin/step-ca
 COPY --from=kms /usr/local/bin/step-kms-plugin /usr/local/bin/step-kms-plugin
 
 USER root
-RUN apk add --no-cache libcap && setcap CAP_NET_BIND_SERVICE=+eip /usr/local/bin/step-ca
-RUN apk add --no-cache pcsc-lite pcsc-lite-libs
+RUN apt-get install -y --no-install-recommends libcap2-bin && \
```
Comment (maraino) [inline docker/Dockerfile.hsm:20]: We might want to add the `apt-get update` here too.

**Review by maraino — CHANGES_REQUESTED** [review]

**Review by maraino — DISMISSED** [review]
Looks good, but the PR in smallstep/step-kms-plugin is not yet approved.

**Comment by tashian** [comment]
@dopey I discussed the image tag naming convention for this with @maraino. Let's do something similar to what [golang](https://hub.docker.com/_/golang) does with tags. I don't think we need _all_ of the tags they use in golang, but a subset:

The following would be alpine builds:
`step-cli:latest`, `step-cli:latest-alpine`, `step-cli:alpine`, `step-cli:0.24.0`, `step-cli:0.24.0-alpine`

for a release candidate: `step-cli:0.24.0-rc.3`, `step-cli:0.24.0-rc.3-alpine`

And the following would be bullseye builds:
`step-cli:latest-bullseye`, `step-cli:bullseye`, `step-cli:0.24.0-bullseye`

for a release candidate: `step-cli:0.24.0-rc.3-bullseye`

File: docker/Dockerfile.hsm [hunk 3]
```diff
@@ -1,23 +1,26 @@
...
 
 COPY --from=builder /src/bin/step-ca /usr/local/bin/step-ca
 COPY --from=kms /usr/local/bin/step-kms-plugin /usr/local/bin/step-kms-plugin
 
 USER root
-RUN apk add --no-cache libcap && setcap CAP_NET_BIND_SERVICE=+eip /usr/local/bin/step-ca
-RUN apk add --no-cache pcsc-lite pcsc-lite-libs
+RUN apt-get update
+RUN apt-get install -y --no-install-recommends libcap2-bin && \
+    setcap CAP_NET_BIND_SERVICE=+eip /usr/local/bin/step-ca
```
Comment (maraino) [inline docker/Dockerfile.hsm:22]: Can we set the capabilities in the builder, and use it in the final image without installing libcap2-bin?

Comment (tashian) [reply docker/Dockerfile.hsm:22]: good idea!

**Comment by tashian** [comment]
@maraino a related thought: If we are doing `step-kms-plugin:cloud` instead of `step-kms-plugin-cloud`, do we also want to push `step-ca:hsm` instead of `step-ca-hsm`?

**Review by maraino — APPROVED** [review]
lgtm

**Review by maraino — APPROVED** [review]

**Pull request merged**